src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -591,3 +646,288 @@ RouterExecutorEnd(QueryDesc *queryDesc)
+		ExpressionWalkerState *parentState)
+{
+	ExpressionWalkerState childState;
+	Node *copy = NULL;
+	bool isBranch = false;
+	bool containsVar = false;
+
+	if (expression == NULL)
+	{
+		return expression;
+	}
+
+	/* pass any argument lists back to the mutator to copy and recurse for us */
+	if (IsA(expression, List))
+	{
+		return expression_tree_mutator(expression,
+									   PartiallyEvaluateExpressionWalker,
+									   parentState);
+	}
+
+	InitializeWalkerState(&childState);
+
+	if (IsA(expression, Var))
+	{
+		containsVar = true;
+	}
+
+	if (IsA(expression, FuncExpr) || IsA(expression, OpExpr))
+	{
+		/* Tell the next step that it should add itself to one of the subtree lists */
+		childState.parentIsBranch = true;
+		isBranch = true;
+	}
+
+	copy = expression_tree_mutator(expression,
+								   PartiallyEvaluateExpressionWalker,
+								   &childState);
+
+	containsVar |= childState.containsVar;
+
+	if (parentState->parentIsBranch)
+	{
+		if (containsVar)
+		{
+			parentState->subtreesWithVar = lappend(parentState->subtreesWithVar, copy);
+			parentState->subtreeOrder = lappend_int(parentState->subtreeOrder, 0);
+		}
+		else
+		{
+			parentState->subtreesWithoutVar = lappend(parentState->subtreesWithoutVar,
+													  copy);
+			parentState->subtreeOrder = lappend_int(parentState->subtreeOrder, 1);
+		}
+	}
+
+	/* never set containsVar=false because parentState is shared between sibling trees */
+	if (containsVar)
+	{
+		parentState->containsVar = true;
+	}
+
+	if (isBranch && parentState->isTopLevel &&
+		list_length(childState.subtreesWithVar) == 0)
+	{
+		return EvaluateExpression(expression);
+	}
+
+	if ((list_length(childState.subtreesWithVar) != 0) &&
+		(list_length(childState.subtreesWithoutVar) != 0))
+	{
+		if (IsA(expression, FuncExpr))
+		{
+			((FuncExpr *) copy)->args = RewriteArgs(&childState);
+		}
+
+		if (IsA(expression, OpExpr))
+		{
+			((OpExpr *) copy)->args = RewriteArgs(&childState);
+		}
+	}
+
+	return copy;
+}
+
+/* evaluate every non-Var arg and return an argument list */
+static List * RewriteArgs(ExpressionWalkerState *state)
+{
+	List *rewrittenArgs = NULL;
+	ListCell *nextArgCell = NULL;
+	int nextArg = 0;
+
+	foreach(nextArgCell, state->subtreeOrder)
+	{
+		nextArg = lfirst_int(nextArgCell);
+		if (nextArg)
+		{
+			Node *original = (Node *)linitial(state->subtreesWithoutVar);
+			Node *evaluated = EvaluateExpression(original);
+			rewrittenArgs = lappend(rewrittenArgs, evaluated);
+			state->subtreesWithoutVar =
+				list_delete_first(state->subtreesWithoutVar);
+		}
+		else
+		{
+			rewrittenArgs = lappend(rewrittenArgs,
+									linitial(state->subtreesWithVar));
+			state->subtreesWithVar = list_delete_first(state->subtreesWithVar);
+		}
+	}
+
+	return rewrittenArgs;
+}
+
+static Node * EvaluateExpression(Node *expression)
+{
+	if(IsA(expression, FuncExpr))
+	{
+		FuncExpr *expr = (FuncExpr *)expression;
+
+		return (Node *) citus_evaluate_expr((Expr *) expr,
```

Comment:
Afaics this will call STRICT functions with NULL arguments, which lead to crashes or worse.  If a function is marked strict, and any of it's arguments are NULL, it'll not be called, but NULL will be returned.  That also exists for OpExprs.